includes/class-fs-garbage-collector.php

```diff
@@ -0,0 +1,359 @@
+<?php
+    /**
+     * @package   Freemius
+     * @copyright Copyright (c) 2015, Freemius, Inc.
+     * @license   https://www.gnu.org/licenses/gpl-3.0.html GNU General Public License Version 3
+     * @since     2.6.0
+     */
+
+    if ( ! defined( 'ABSPATH' ) ) {
+        exit;
+    }
+
+    class FS_Garbage_Collector {
+        #--------------------------------------------------------------------------------
+        #region Singleton
+        #--------------------------------------------------------------------------------
+
+        private $products_to_skip_by_type_and_slug;
+
+        /**
+         * @var FS_Garbage_Collector
+         * @since 2.6.0
+         */
+        private static $_instance;
+
+        /**
+         * @return FS_Garbage_Collector
+         */
+        static function instance() {
+            if ( ! isset( self::$_instance ) ) {
+                self::$_instance = new self();
+            }
+
+            return self::$_instance;
+        }
+
+        #endregion
+
+        private function __construct() {
+        }
+
+        function run() {
+            require_once WP_FS__DIR_INCLUDES . '/class-fs-lock.php';
+
+            $lock = new FS_Lock( 'garbage_collection' );
+
+            if ( $lock->is_locked() ) {
+                return;
+            }
+
+            // Create a 1-day lock.
+            $lock->lock( WP_FS__TIME_24_HOURS_IN_SEC );
+
+            $this->clean_up();
+       }
+
+        private function clean_up() {
+            $_accounts = FS_Options::instance( WP_FS__ACCOUNTS_OPTION_NAME, true );
+
+            $products_to_delete = array();
+
+            foreach ( $this->get_product_types() as $product_type ) {
+                $products_to_delete[ $product_type ] = $this->get_products_to_delete( $_accounts, $product_type );
+            }
+
+            $update_options = false;
+
+            if ( ! empty( $products_to_delete ) ) {

```

Comment:
@fajardoleo Continued from below...

If we do it like this

```php
if ( empty( $products_to_delete ) ) {
  return;
}
```

Then we won't need to hold a flag and compromise on performance. Plus I think the code would be cleaner.